### PR TITLE
[HOLY SHIT DO NOT MERGE THIS] Abrtracts species to DNA

### PR DESCRIPTION
**Pull request by Aurorablade**
This is an attempt to make species non global..so i don;t accidently change one persons icon and make everyone spiders again.

Its also just better.

Now obviously at the moment this code is rougher then [lewd here] with a [lewd here] while [lewd here] and a flower.

i got runtimes https://hastebin.com/coxugosupu.sql

so why am i making a PR, I like extra set of eyes on my code telling what i need to change and what i am doing wrong. When you work in retail you sometimes come home and are just brain effed.

In the process of making this compile i heard the voice of cthulhu about 20 times.
Added things (i may be missing some):

 var/id      // if the game needs to manually check your race to do something not included in a proc here, it will use this
 var/meat = /obj/item/weapon/reagent_containers/food/snacks/meat/human //What the species drops on gibbing
 var/skinned_type
 var/nojumpsuit = 0 // this is sorta... weird. it basically lets you equip stuff that usually needs jumpsuits without one, like belts and pockets and ids
 var/attack_verb = "punch" // punch-specific attack verb
 var/sound/attack_sound = 'sound/weapons/punch1.ogg'
 var/sound/miss_sound = 'sound/weapons/punchmiss.ogg'
 var/list/no_equip = list() // slots the race can't equip stuff to

 
procs:
//Called when cloning, copies some vars that should be kept
/datum/species/proc/copy_properties_from(datum/species/old_species)
 return
 
On_species_loss/gain

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
somehow, this actually compiles.

**Comment by Aurorablade** [comment]
After some work yes. Now I have to figure out the runtimes

**Comment by KasparoVy** [comment]
I think I have an idea about what might be causing the runtimes. Will take a look when i get home...  Eventually.

**Comment by Aurorablade** [comment]
@KasparoVy are you performing icon voodoo over there again?

**Comment by KasparoVy** [comment]
Undid a bunch of my voodoo actually now that it's no longer necessary. I've narrowed it down to one runtime/bug that causes the mob's organs not to spawn, but I got stumped there last night. I'll show you where I'm at when I get home and keep trying to sort that last bit out

**Comment by KasparoVy** [comment]
Done @Aurorablade, see [this commit](https://github.com/KasparoVy/Paradise/commit/37ab403e6546cb3a92a123fe38e3081c984f3593)
A lot of brain-busting was involved, I assure you.

**Comment by Aurorablade** [comment]
will look at it a bit.

**Comment by KasparoVy** [comment]
If it looks good enough and you want to directly merge the changes, cherry-pick's the command
It's what I had to do to check out your commits haha

**Comment by Aurorablade** [comment]
@KasparoVy give me the command to cherry-pick this or make a merge request for it on my branch in my rp. I am trying but having to trouble.

**Comment by KasparoVy** [comment]
Roger roger. Here goes.

**Comment by Aurorablade** [comment]
finding some more runtimes here
Runtime in preferences_setup.dm,4: Cannot create objects of type /list.

 [18:22:48] Runtime in organ_icon.dm,54: Cannot read /datum/species/machine (/datum/species/machine).name

trying to sort em out.

**Comment by Aurorablade** [comment]
[18:22:48] Runtime in organ_icon.dm,54: Cannot read /datum/species/machine (/datum/species/machine).name
  proc name: get icon (/obj/item/organ/external/proc/get_icon)
  usr: Monte Burkett (fethas) (/mob/living/carbon/human)
  usr.loc: The floor (49,138,1) (/turf/simulated/floor/plasteel)
  src: Morpheus Cyberkinetics upper b... (/obj/item/organ/external/chest/ipc)
  src.loc: Monte Burkett (/mob/living/carbon/human)
  call stack:
  Morpheus Cyberkinetics upper b... (/obj/item/organ/external/chest/ipc): get icon(null, null)
  Morpheus Cyberkinetics upper b... (/obj/item/organ/external/chest/ipc): robotize("Morpheus Cyberkinetics", 0, 1)
  Morpheus Cyberkinetics upper b... (/obj/item/organ/external/chest/ipc): New(Monte Burkett (/mob/living/carbon/human))
  Machine (/datum/species/machine): create organs(Monte Burkett (/mob/living/carbon/human), null)
  Machine (/datum/species/machine): on species gain(Monte Burkett (/mob/living/carbon/human), null)
  Monte Burkett (/mob/living/carbon/human): set species(Machine (/datum/species/machine), null, 1)
  Monte Burkett (/mob/living/carbon/human): change species("Machine")
  Appearance Editor (/datum/nano_module/appearance_changer): Topic("src=\[0x21005ff4]_129;race=Mac...", /list (/list), 0, /datum/topic_state/default (/datum/topic_state/default))
  /datum/nanoui (/datum/nanoui): Topic("src=\[0x21005ff4]_129;race=Mac...", /list (/list))
  Fethas (/client): Topic("src=\[0x21005ff4]_129;race=Mac...", /list (/list), /datum/nanoui (/datum/nanoui))

Also sometimes the race change useing the Admin CMA panel won't take icon wise until i regenerate icons by force via VV.

**Comment by Aurorablade** [comment]
@KasparoVy remind me to add you to the changelog.
I know alot of this is cuase we need some better icon code ala tg or somewhere else.

**Comment by Aurorablade** [comment]
Okay i may have fixed the one thing about the list but heres a gif of what i am experianceing with the icon changes:
https://puu.sh/zncIY/a5dfc3ae2a.gif

**Comment by KasparoVy** [comment]
Lmao, that's an interesting bug. It's definitely 'cause of some wack runtime that I seem to've missed. I'll look through this some more

**Comment by Aurorablade** [comment]
@KasparoVy i think the fact theres a species datum call in organ.dm  (theres one thats a list in gland.dm) may be cuaseing some confusion on var set up at least from code reading stand point, it may need to be changed, but some of this may be icon related. If you can figure something out it will help we may be able to remove that species datum var in organ.dm.

**Comment by KasparoVy** [comment]
@Aurorablade the species datum var exists in the event the lomb os transplanted onto another person or is spawned without a holder and needs a species so it can still behave as intended (and in the case of heads, produce only the intended hairstyles) 

Will try to make tine to look at this tonight

**Comment by Aurorablade** [comment]
@KasparoVy no rush its just kinda coder thoughts i share, may just need to rename it for readability.

**Comment by KasparoVy** [comment]
Lmao, this one is hilarious. Just checking it out now. This'll be fun to debug.
Yeah, will be fine. I just realized, too, we may not need it either if we refactor the code to read species off limb DNA just safety it like what's done for the mobs.

**Comment by Aurorablade** [comment]
again and its probaly another PR we could use some icon code updates. I know tg has stuff for limb generation and stuff, but thats your niche, icon code is voodoo to me. (our icon code sucks plz help)

**Comment by KasparoVy** [comment]
Fixed the species-changing bug, just trying to sort that machine runtime now.

**Comment by KasparoVy** [comment]
@Aurorablade All fixed, see [this commit](https://github.com/KasparoVy/Paradise/commit/70882dead7ffce93d0cc4682cd0fc03b51e7bd5b)! The species-changing and IPC runtimes are resolved.
I did additional testing where-by I grafted a Tajaran's arm onto a Vulpkanin and found a bug there as well that I fixed with the above commit.

The previous commit titled [maybe a fix](https://github.com/ParadiseSS13/Paradise/pull/8656/commits/2e642c77f3378ed4e60ee721ea71eaeffae83612) is not necessary and can be reverted.

**Comment by Aurorablade** [comment]
Will look after work but this also proves coders make Eldritch horrors to test code on a regular basis

**Comment by KasparoVy** [comment]
Summoning code'sie.

**Comment by Aurorablade** [comment]
@KasparoVy that fixed it but i got to figure out this now [15:49:15] Runtime in preferences_setup.dm,4: Cannot create objects of type /list.
  proc name: random character (/datum/preferences/proc/random_character)
  usr: Fethas (fethas) (/mob/new_player)
  usr.loc: The floor (8,174,2) (/turf/unsimulated/floor)
  src: /datum/preferences (/datum/preferences)
  call stack:
  /datum/preferences (/datum/preferences): random character(null)
  /datum/preferences (/datum/preferences): process link(Fethas (/mob/new_player), /list (/list))
  Fethas (/client): Topic("_src_=prefs;preference=all;tas...", /list (/list), 

thats on char gen. it won't random, i mean before the found starts . i need to see if it does i after start

**Comment by Aurorablade** [comment]
yeah it does

**Comment by Aurorablade** [comment]
it works if i take the new off it.

**Comment by Aurorablade** [comment]
huh so theres a bug, also on live, that when you change species it drops like shoes and headset.
I see why, its cause create organs is replacing all the external parts, and you do that it calls the organs remove proc which drops the organs worn items.

File: code/game/machinery/computer/cloning.dm [hunk 1]
```diff
@@ -405,7 +405,7 @@
 		var/datum/species/S = all_species[R.dna.species]
```
Comment (KasparoVy) [inline code/game/machinery/computer/cloning.dm]: Hmmm. This 'might' need to be all_species[R.dna.species.name].

File: code/modules/mob/new_player/preferences_setup.dm [hunk 2]
```diff
@@ -1,10 +1,10 @@
 /datum/preferences
 	//The mob should have a gender you want before running this proc. Will run fine without H
 /datum/preferences/proc/random_character(gender_override)
-	var/datum/species/S = new all_species[species]
+	var/datum/species/S = all_species[species]
```
Comment (KasparoVy) [inline code/modules/mob/new_player/preferences_setup.dm]: Totally right, this species var being a unique type isn't necessary for this proc. Silly me

**Comment by KasparoVy** [comment]
Lmao, I notice that one too. Pretty funny, but not much we could/should be doing about it here.

**Comment by Aurorablade** [comment]
well at least i confirmed its not just my code...i am wondering if could put it on the drop organs proc i ported over..

**Comment by KasparoVy** [comment]
Maybe we could have it up in the `change_species()` proc in `appearance.dm`.
Like, we cache a list of all the mob's equipment, do the species stuff, then at the very end attempt to re-equip it all if it isn't already. *shrug*

I think it's way more trouble than it's worth though.

**Comment by KasparoVy** [comment]
Nice, good find 😁

**Comment by Aurorablade** [comment]
Closing as Fox has it covered..and i did inspire a few things over there!

**Pull request closed without merging**